src/compiler/transformers/rewrite-aliased-paths.ts

```diff
@@ -0,0 +1,152 @@
+ * {@see https://github.com/microsoft/TypeScript/issues/10866} and have
+ * said that resolving these module identifiers is the responsability of module
+ * bundling and build tools.
+ *
+ * So that means we've got to do it!
+ *
+ * This visitor function does so by getting the resolved file path to any
+ * module which is not 1) not external (i.e. not a dependency) and 2) is not
+ * already a relative, file-path based import. It then replaces the module
+ * identifier with the relative path from the importer to the importee.
+ *
+ * @param compilerHost a TS compiler host
+ * @param transformCtx a TS transformation context
+ * @param sourceFilePath the path to the source file being visited
+ * @returns a visitor which takes a node and optionally transforms imports
+ */
+const visit = (compilerHost: ts.CompilerHost, transformCtx: ts.TransformationContext, sourceFilePath: string) => {
+  return (node: ts.Node): ts.VisitResult<ts.Node> => {
+    if (ts.isImportDeclaration(node) && ts.isStringLiteral(node.moduleSpecifier)) {
+      let importPath = node.moduleSpecifier.text;
+
+      // We will ignore transforming any paths that are already relative paths or
+      // imports from external modules/packages
+      if (!importPath.startsWith('.')) {
+        const module = ts.resolveModuleName(
+          importPath,
+          sourceFilePath,
+          transformCtx.getCompilerOptions(),
+          compilerHost
+        );
+
+        const hasResolvedFileName = module.resolvedModule?.resolvedFileName != null;
+        const isModuleFromNodeModules = module.resolvedModule?.isExternalLibraryImport === true;
+        const shouldTranspileImportPath = hasResolvedFileName && !isModuleFromNodeModules;
+
+        if (shouldTranspileImportPath) {
+          // Create a regular expression that will be used to remove the last file extension
+          // from the import path
+          const extensionRegex = new RegExp(
+            Object.values(ts.Extension)
+              .map((extension) => `${extension}$`)
+              .join('|')
+          );
+
+          // In order to make sure the relative path works when the destination depth is different than the source
+          // file structure depth, we need to determine where the resolved file exists relative to the destination directory
+          const resolvePathInDestination = module.resolvedModule.resolvedFileName;
+
+          importPath = normalizePath(
+            relative(dirname(sourceFilePath), resolvePathInDestination).replace(extensionRegex, '')
+          );
+        }
+      }
+
+      return transformCtx.factory.updateImportDeclaration(
+        node,
+        retrieveTsModifiers(node),
+        node.importClause,
+        transformCtx.factory.createStringLiteral(importPath),
+        node.assertClause
+      );
+    }
+
+    return ts.visitEachChild(node, visit(compilerHost, transformCtx, sourceFilePath), transformCtx);
```

Comment:
I poked around a bit and found that the following (contrived) example does seem to be valid TypeScript:

```ts
namespace Foo {
    export function foobar () {
        return 2;
    }
}

namespace Bar {
    import foo = Foo;

    export const baz = foo.foobar;
}
```

([TS playground link](https://www.typescriptlang.org/play?#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuSAZgK7DwUCWawNGARiGUgBQCU2PARFkIFWmTYAmANzD8AXxzKcoSLATIAQtyEimYUpXaYAvKgzyFhEuSrxWUKlwBeSC9QwA6L2i5k8opAA) for that snippet)

_However_, if you take a look at [the AST for that snippet](https://ts-ast-viewer.com/#code/HYQwtgpgzgDiDGEAEAxA9mpBvAUE-SEAHjGgE4AuSAZgK7DwUCWawNGARiGUgBQCU2PARFkIFWmTYAmANzD8AXxzKcoSLATIAQtyEimYUpXaYAvKgzyFhEuSrxWUKlwBeSC9QwA6L2i5k8opAA), happily the `import foo = Foo;` bit is not a `ts.ImportDeclaration`, but a `ts.ImportEqualsDeclaration`, and I don't think that TypeScript will allow a normal import declaration within a namespace (I could be wrong though). So that would suggest we could get by with filtering on the top-level statements in a `ts.SourceFile`.

<img width="390" alt="Screenshot 2023-02-15 at 2 30 53 PM" src="https://user-images.githubusercontent.com/6207644/219133031-9d375e5c-9eeb-4c15-8492-2349e0e3dab7.png">

In the future, however, if the [module declarations proposal](https://github.com/tc39/proposal-module-declarations) get accepted as part of the ECMAScript standard (it's [currently stage 2](https://github.com/tc39/proposals#stage-2)) then I think it would be possible to have something like this:

```ts
module Foo {
  import fs from 'fs';
  export const myFileSync = fs.readFileSync;
}
```

so in the interests of future-proofing we might want to keep the recursion around.

Thoughts?

Reply:
I struggle with this decision (or, my opinion on it 😆). IIUC, the module declarations proposal just hit Stage 2 this past November - I think I'd err on the side of [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) - and if we _do_ need it in the future:
1. This probably won't be the only change we'll need to make, and I can see value in looking at the solution for module declarations holistically
2. We have git history, and can recover this original impl and reuse it if needed

Reply:
I think that makes sense, it's not going to be that difficult retrofit recursion back in if we turn out to need it.

I'm going to finish tests for the current implementation first and then look at swapping things out, just to have some assurance that things are working as I think they are.

Reply:
alright so I just pushed a change that did a bit of a refactor on the `rewrite-aliased-paths` module to avoid recurring down the whole tree, lmk what you think!